### ISSUE TITLE
voluptuous.error.MultipleInvalid: extra keys not allowed @ data['icon']

### ISSUE THREAD

**Issue by jacquesvzijp** [harriedegroot/nl.hdg.mqtt#33]
Hi Harrie,
Thanks for the great work. Below an error message I get for certain KAKU dimmer devices. HA seems not to like the icon that is passed by the Hub.
Thanks.

Home Assistant 0.111.3

voluptuous.error.MultipleInvalid: extra keys not allowed @ data['icon']

2020-06-24 15:57:19 ERROR (MainThread) [homeassistant.util.logging] Exception in async_discover when dispatching 'mqtt_discovery_new_light_mqtt': ({'name': 'Achtergrondlicht', 'unique_id': '143e2315-051f-487d-8f65-b69c3a438e09_light', 'payload_on': 'true', 'payload_off': 'false', 'state_topic': 'homie/homey-570bafc09aac671e67accde0/achtergrondlicht/onoff', 'state_value_template': '{{ value }}', 'command_topic': 'homie/homey-570bafc09aac671e67accde0/achtergrondlicht/onoff/set', 'on_command_type': 'brightness', 'icon': 'mdi:toggle-switch-off-outline', 'brightness_state_topic': 'homie/homey-570bafc09aac671e67accde0/achtergrondlicht/dim', 'brightness_command_topic': 'homie/homey-570bafc09aac671e67accde0/achtergrondlicht/dim/set', 'brightness_value_template': '{{ value }}', 'brightness_scale': 100, 'device': {'identifiers': 'homey-570bafc09aac671e67accde0_143e2315-051f-487d-8f65-b69c3a438e09', 'name': 'Achtergrondlicht'}, 'platform': 'mqtt'},)

**Comment by jacquesvzijp** [comment]
After some struggling I managed to manually add these entities in my configuration.yaml, leaving out the icon attribute. Now it is all fine. Thanks again for the great work!

**Comment by harriedegroot** [comment]
Great you got it working!

Do you know if something changed in the HA mqtt discovery check? Does this problem exists for all devices, or just the KAKU ones? In other words, should I completely remove the icon property from the hub? Just in specific cases, or leave it just as it is?

**Comment by jacquesvzijp** [comment]
Harry,
I found same behavior for a Tuya lamp. With MQTT explorer I could see all devices with icon attribute which gave this error in the HA configuration. I also have Fibaro, NEO and Aetoc devices that do not give the problem (but these do not have the icon). 
I only recently installed HA so cannot say if this changed from previous versions but I do find others (https://community.home-assistant.io/t/custom-icon-in-mqtt-switch/1267) having that issue. It seems icons can only be added in the front end.
So it seems best to remove icon altogether.
Thanks
Jacques

**Comment by rindlerblabla** [comment]
> Great you got it working!
> 
> Do you know if something changed in the HA mqtt discovery check? Does this problem exists for all devices, or just the KAKU ones? In other words, should I completely remove the icon property from the hub? Just in specific cases, or leave it just as it is?

Think you should remove the icon property. Would also solve #45 and https://github.com/harriedegroot/nl.hdg.mqtt/issues/41#issuecomment-705349118 . Until then it's possible to just copy the json phrase from the error message after removing the icon property to eg. homeassistant/switch/name/config according to https://www.home-assistant.io/docs/mqtt/discovery/#switches

**Comment by Pantsema** [comment]
Hi all, since I have a couple of different KAKU devices connected to Homey and also experienced this issue I'd like to add some information about my discoveries...

First of all, most devices are discovered by Homeassistant without problem. This specific problem only occurs to me on the KAKU ACM-100 (built-in dimmer) and ACD (socket dimmer). These device types seem to send the extra icon attribute which causes HA to trow an error (shouldn't they just ignore the extra value and trow a warning??). What I've found out that these devices in their homie/homey{yourtags}/device have an attribute $type= set to socket. There seem to be different sources of device types because these mentioned devices are published in the discovery topics homeassistant/light in stead of the homeassistant/switch (where the on/off KAKU devices are posted). Within the last topic there is a structure of homeassistant/switch/{name}/onoff and the config message is published within that channel (and does have a working icon attribute and no state_value_template (see below).

The workaround was to manually add the devices to the configuration.yaml or publish the message without icon attribute to the topic. Both essentially work but don't stop the errors getting into your logs every HA reboot (and that happens a lot). Also when manually helping the discovery proces by publishing the message I also had to remote the "state_value_template" otherwise the state would to change in HA for some reason.

Not sure what the exact fix needs to right now.. 

1. Scenario 1; remove the icon attribute; will make sure the discovery finishes, but will have collateral damage to devices with a working icon and also probably need to remove the state_value_template as well??
2. Scenario 2: correctly broadcast the devices with their types;
- A) either these are lights as per /homeassistant/ topic and then the type need to reflect light as well and then the homie/homey topics need the device type to reflect light as well.
- B) as per homie/homey topic these are sockets and then need to broadcast in the homeassistant topic as ... switch? with dim properties.. 
- C).. my favorite :)... use the device type as set up by the user in homey. I've set up these sockets, but also the on/off devices as lights (as per homey's question; whats been plugged in?). I'd like to have these types synced to homeassistant as well (for the on/off devices I've now manually add a light device in homeassistant.

Long story, sorry for taking your time, but hope this helps. I've any additional questions arise or you'd like me to test anything, let me know!
